### PR TITLE
feat(log): request/response logging decorator

### PR DESCRIPTION
**Pull request by stikos**
## Description

Creation of a configurable decorator to log requests and their responses on the desired log level

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist:

- [x] My code follows the [style guidelines](https://github.com/elixir-cloud-aai/elixir-cloud-aai/blob/dev/resources/contributing_guidelines.md#language-specific-guidelines) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

### CODE REVIEW

File: foca/utils/logging.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import logging
+from connexion import request
+from functools import wraps
+from typing import (Callable, Optional, )
+
+logger = logging.getLogger(__name__)
+
+
+def log_traffic(_fn: Callable = None,
+                log_request: Optional[bool] = True,
```
Comment (uniqueg) [inline foca/utils/logging.py]: I think you may have misunderstood the meaning of `Optional[]`. It doesn't have anything to do whether an argument is required - it says that the _type_ of the argument can be either `None` or the type indicated in between the angular brackets.

In fact, the only argument that has an `Optional[]` type is `_fn` (which can be a callable or `None`) - whereas all the others are _always_ either `bool` (`log_request` and `log_response`) or `int` (`log_level`).

To be more explicit, I have changed the `log_level` default from `20` to `logging.INFO`.

File: foca/utils/logging.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                log_response: Optional[bool] = True,
+                log_level: Optional[int] = 20,
+                ) -> Callable:
+    """Customizable decorator for logging requests and responses on
```
Comment (uniqueg) [inline foca/utils/logging.py]: No need to point out that a function is customizable - this is obvious from the argument section. Similarly, no need to point out arguments explicitly in the summary line.

File: foca/utils/logging.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    configurable log level.
+
+    Args:
+        _fn: The function to be decorated (gets passed implicitly).
```
Comment (uniqueg) [inline foca/utils/logging.py]: Private arguments don't need to be described in the API docs.

File: foca/utils/logging.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        log_request: A boolean flag to set whether or not the request should
```
Comment (uniqueg) [inline foca/utils/logging.py]: No need to mention types and defaults, this is obvious from the function's signature.

File: foca/utils/logging.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                               f"{request.environ['PATH_INFO']} "
+                               f"{request.environ['SERVER_PROTOCOL']}\" from "
+                               f"{request.environ['REMOTE_ADDR']}"
+                           )
+
+            response = fn(*args, **kwargs)
+            if log_response:
+                logger.log(level=log_level,
+                           msg=f"Response to request "
+                               f"\"{request.environ['REQUEST_METHOD']} "
```
Comment (uniqueg) [inline foca/utils/logging.py]: Unnecessary repetition. Write the request to a variable and use it in both the log request and log response sections.

File: foca/utils/logging.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+"""Decorator for logging requests and responses."""
```
Comment (uniqueg) [inline foca/utils/logging.py]: There may be more logging utilities added to this module in the future, so better stay generic here. The decorator is described in its own docstring.

File: tests/utils/test_logging.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+"""Tests for the logging decorator"""
```
Comment (uniqueg) [inline tests/utils/test_logging.py]: Again - there may be tests added for other functions in the future.

File: tests/utils/test_logging.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+       'PATH_INFO': '/',
+       'SERVER_PROTOCOL': 'HTTP/1.1',
+       'REMOTE_ADDR': '192.168.1.1',
+       }
+
+# Get logger instance
+logger = logging.getLogger(__name__)
+
+
+def test_log_level(caplog):
```
Comment (uniqueg) [inline tests/utils/test_logging.py]: Better to include what you are testing in case other functions are added.

File: tests/utils/test_logging.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
+"""Tests for the logging decorator"""
+
+import logging
```
Comment (uniqueg) [inline tests/utils/test_logging.py:3]: Check how imports should be ordered in Python. See [here](https://www.python.org/dev/peps/pep-0008/#imports) for the relevant section in PEP8.

**Review by uniqueg — APPROVED** [review]
@stikos Looks great, just minor details. I've done the changes already, comments are just FYI

File: foca/utils/logging.py (hunk 1 above)
Comment (stikos) [reply foca/utils/logging.py]: I definitely had misunderstood it; thanks for the clarification

**Pull request merged**